### PR TITLE
[core] Fix #2639, create report file directories if they do not exist

### LINKED ISSUE #2639: [core] PMD CLI output file is not created if directory or directories in path don't exist
**Affects PMD Version:** 
6.25.0

**Description:**
The option -r which specifies the report file throws an exception if directory or directories in the specified report path do not exist.

**Exception StackTrace**

FINE: Exception during processing           java.lang.IllegalArgumentException: java.nio.file.NoSuchFileException: /data/data/com.termux/files/home/DSAlgos/target/pmdxml.txt           at net.sourceforge.pmd.util.IOUtil.createWriter(IOUtil.java:43)                         at net.sourceforge.pmd.PMD.doPMD(PMD.java:221)                                          at net.sourceforge.pmd.PMD.run(PMD.java:482)                                            at net.sourceforge.pmd.cli.PMDCommandLineInterface.run(PMDCommandLineInterface.java:166)                                            at net.sourceforge.pmd.PMD.main(PMD.java:452)                                   Caused by: java.nio.file.NoSuchFileException: /data/data/com.termux/files/home/DSAlgos/target/pmdxml.txt                                    at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)                                                 at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)                                                  at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)                                                  at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)                                      at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)                                      at java.base/java.nio.file.Files.newOutputStream(Files.java:219)                        at java.base/java.nio.file.Files.newBufferedWriter(Files.java:2919)                     at net.sourceforge.pmd.util.IOUtil.createWriter(IOUtil.java:41)                         ... 4 more

**Running PMD through:** *[CLI ]*

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

(The bug was that the parent directories of the report file were not created. The file itself was already properly created.)

Fixes #2639 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/380aef51ba4e405dbcbee96ddff2c158/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/test/java/net/sourceforge/pmd/cli/CoreCliTest.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        assertTrue("Report file should exist", Files.exists(reportFile));
+
+        runPmdSuccessfully("-d", srcDir, "-R", DUMMY_RULESET, "-r", reportFile);
+
+        assertNotEquals(readString(reportFile), STRING_TO_REPLACE);
+    }
+
+    @Test
+    public void testNonExistentReportFile() {
+        Path reportFile = tempRoot().resolve("out/reportFile.txt");
```
Comment (adangel) [inline pmd-core/src/test/java/net/sourceforge/pmd/cli/CoreCliTest.java:68]: I'd use here a different intermediate directory. If this test runs after the other, then we don't test, whether the directory is created... or we need another test case "testNonExistentReportDirAndFile()"

**Review by adangel — APPROVED** [review]
I'll add another test case and merge this. Thanks!

**Comment by adangel** [comment]
> I'll add another test case and merge this. Thanks!

Ah, right - the test rules are executed for every unit test - so the directory doesn't actually exist.... ok, that's good.

**Pull request merged**